Avoid scenarios where every element automatically gets an interaction region

etiennesegonzac opened the pull request:
#### fbf81751af85bcc9401029fd025e51ed03dfa91c
<pre>
Avoid scenarios where every element automatically gets an interaction region
<a href="https://bugs.webkit.org/show_bug.cgi?id=255220">https://bugs.webkit.org/show_bug.cgi?id=255220</a>
&lt;rdar://100421682&gt;

Reviewed by Tim Horton.

We have a corner case where every element on the page gets an
interaction region regardless of its own properties: a body element that
both does event delegation _and_ sets `cursor: pointer;`.
Work around this issue by ignoring the `pointer` in this specific case.

* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):
Detect the corner case and ignore `pointer` on the body.

* LayoutTests/interaction-region/ignore-catch-all-body-expected.txt: Added.
* LayoutTests/interaction-region/ignore-catch-all-body.html: Added.
Add a test case.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(dumpCALayer):
* LayoutTests/interaction-region/layer-tree-expected.txt:
Fix test flakiness by not dumping InteractionRegion layers&apos; sublayers.

Canonical link: <a href="https://commits.webkit.org/263077@main">https://commits.webkit.org/263077@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d8283af7d6c3bb4a9084b66be10a83d86a28df22

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3276 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3334 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3445 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4686 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3603 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3246 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3391 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3355 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2840 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3310 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3601 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2942 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4508 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1103 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2912 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2833 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2884 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2965 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4246 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3345 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2671 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2911 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2913 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/860 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2907 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3186 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebCore/style/StyleAdjuster.cpp:
```diff
@@ -583,6 +583,12 @@ void Adjuster::adjust(RenderStyle& style, const RenderStyle* userAgentAppearance
+#if ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)
+        // Every element will automatically get an interaction region which is not useful, ignoring the `cursor: pointer;` on the body.
+        if (is<HTMLBodyElement>(*m_element) && style.cursor() == CursorType::Pointer && style.eventListenerRegionTypes().contains(EventListenerRegionType::MouseClick))
+            style.setCursor(CursorType::Auto);
```
I'm not sure we should be *actually* adjusting the style like this, should we? Is it possible to either identify this later (in InteractionRegion), to keep the heuristic all in one place, or to just remove the bit from `eventListenerRegionTypes` instead of actually mutating the style in a visible way?

hortont424 replied:
We can go with this for now since nobody came up with a better plan.

etiennesegonzac committed:
Avoid scenarios where every element automatically gets an interaction region
https://bugs.webkit.org/show_bug.cgi?id=255220
<rdar://100421682>

Reviewed by Tim Horton.

We have a corner case where every element on the page gets an
interaction region regardless of its own properties: a body element that
both does event delegation _and_ sets `cursor: pointer;`.
Work around this issue by ignoring the `pointer` in this specific case.

* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):
Detect the corner case and ignore `pointer` on the body.

* LayoutTests/interaction-region/ignore-catch-all-body-expected.txt: Added.
* LayoutTests/interaction-region/ignore-catch-all-body.html: Added.
Add a test case.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(dumpCALayer):
* LayoutTests/interaction-region/layer-tree-expected.txt:
Fix test flakiness by not dumping InteractionRegion layers' sublayers.

Canonical link: https://commits.webkit.org/263077@main
